File: stack/c/stack.c
#include "stdlib.h"
#include "stdio.h"


typedef struct Node {
  int value;
  struct Node* next;
} Node;

typedef struct Stack {
  int size;
  Node* head;
} Stack;

Stack createStack(){
  Stack stack = { .head = NULL, .size=0 };
  return stack;
}

void push(Stack* stack, int item){
  Node* elem = (Node*) malloc(sizeof(Node));
  Node* prev_head = stack->head;

  stack->size++;
  elem->value = item;
  elem->next = prev_head;

  stack->head = elem;
}

int pop(Stack* stack){

  if(!stack->head){ return -1; }

  Node* elem = stack->head;

  stack->head = elem->next;
  stack->size--;

  int value = elem->value;
  free(elem);

  return value;
}

int peek(Stack* stack){
  return stack->head->value;
}

void printStack(Stack* stack){
  printf("Stack of size %d\n", stack->size);
  Node *curr = stack->head;
  while(curr){
    printf("-> %d\n", curr->value);
    curr = curr->next;
  }
}

int main(){

  Stack stack = createStack();

  push(&stack, 1);
  push(&stack, 2);
  push(&stack, 3);
  printStack(&stack);

  printf("peek: %d\n", peek(&stack));
  printf("pop: %d\n", pop(&stack));
  printf("pop: %d\n", pop(&stack));
  printStack(&stack);

  return 0;
}

File: stack/ts/stack.ts

class StackNode<T>{
    public next: StackNode<T> = null
    constructor(public value: T){}
}

class Stack<T>{
    private size: number = 0
    private head: StackNode<T> = null
    constructor(){}

    get length(): number {
        return this.size
    }

    public push(value: T): void {
        const node = new StackNode(value)
        const prev_head = this.head

        node.next = prev_head
        this.head = node
        this.size++
    }

    public pop(): T {
        if(!this.head){
            return null
        }

        const prev_head = this.head
        this.head = prev_head.next
        this.size--
        return prev_head.value
    }

    public peek(): T {
        return this.head.value
    }

    public print(): void{
        console.log("Stack of size %d", this.size)
        let curr = this.head
        while (curr) {
            console.log("->", curr.value)
            curr = curr.next
        }
    }

}

const stack = new Stack<number>()
stack.push(1)
stack.push(2)
stack.push(3)

stack.print()

console.log("peek:", stack.peek())

console.log("pop:", stack.pop())
console.log("pop:", stack.pop())

stack.print()
